File: BOJ/Python/1005-ACM Craft.py
import sys
from collections import deque
input = sys.stdin.readline
queue = deque()
for _ in range(int(input())):
    n,k = map(int, input().split())
    times = [int(i) for i in input().split()]
    tmp = [True]*n
    graph = [None]*n
    CNT = 0
    for i in range(k):
        a,b = map(int, input().split())
        if graph[a-1]:
            graph[a-1].append(b-1)
        else:
            graph[a-1] = [b-1]
        tmp[b-1] = False
    w = int(input())
    for i in graph:
        if i == None:  continue
        CNT += i.count(w-1)
    queue.clear()
    for i in range(len(tmp)):
        if tmp[i]:
            queue.append((i, times[i]))
    max_ = 0
    c = 0
    while True:
        if not len(queue) or CNT < c:   break
        x,t = queue.popleft()
        if x == w-1:
            if max_ < t:   max_ = t
            c += 1
            continue
        tmp = []
        if graph[x] == None:  continue
        for i in graph[x]:
            tmp.append(times[i])
        t += max(tmp)
        for j in graph[x]:
            queue.append((j,t))
    print(max_)

File: BOJ/C++/1005-ACM Craft.cpp
#include <cstdio>
#include <vector>
#include <queue>

int T, n, k, x, y, w, d[1010], price, a;
std::vector<int> graph[1010];
std::queue<int> queue;
bool calced[1010];

int max(int x, int y) {
    return (x>y)?x:y;
}


int main() {
    scanf("%d", &T);
    while(T--) {
        scanf("%d %d", &n, &k);
        for(int i = 0; i < n; i++) scanf("%d", &d[i]);
        for(int i = 0; i < 1010; i++) graph[i].clear();
        for(int i = 0; i < 1010; i++) calced[i] = false;
        for(int i = 0; i < queue.size(); i++) queue.pop();
        for(int i = 0; i < k; i++) {
            scanf("%d %d", &x, &y);
            graph[y].push_back(x);
        }
        scanf("%d", &w);
        price = d[w-1];
        queue.push(w);
        while(queue.size() != 0) {
            a = 0;w 
            for(int i = 0; i < queue.size(); i++) {
                a = max(a, queue.front());
                queue.pop();

            }
        }
        printf("%d\n", price);
    }
}